python/ray/serve/controller.py

```diff
@@ -276,55 +487,51 @@ async def _recover_from_checkpoint(self, checkpoint_bytes: bytes) -> None:
         start = time.time()
         logger.info("Recovering from checkpoint")
 
-        # Load internal state from the checkpoint data.
-        (
-            self.routes,
-            router_node_ids,
-            self.backends,
-            self.traffic_policies,
-            self.replicas,
-            self.replicas_to_start,
-            self.replicas_to_stop,
-            self.backends_to_remove,
-            self.endpoints_to_remove,
-        ) = pickle.loads(checkpoint_bytes)
-
-        for node_id in router_node_ids:
+        restored_checkpoint: Checkpoint = pickle.loads(checkpoint_bytes)
+        # Restore ConfigurationStore
+        self.configuration_store = restored_checkpoint.config
+
+        # Restore ActorNursery
+        self.actor_nursery = restored_checkpoint.nursery
+
+        # Refresh the RouterCache
+        for node_id in self.actor_nursery.routers_cache.keys():
             router_name = format_actor_name(SERVE_PROXY_NAME,
                                             self.controller_name, node_id)
-            self.routers[node_id] = ray.get_actor(router_name)
-
+            self.actor_nursery.routers_cache[node_id] = ray.get_actor(
+                router_name)
         # Fetch actor handles for all of the backend replicas in the system.
         # All of these workers are guaranteed to already exist because they
-        # would not be written to a checkpoint in self.workers until they
-        # were created.
-        for backend_tag, replica_tags in self.replicas.items():
+        # would not be written to a checkpoint in self.actor_nursery.workers
+        # until they were created.
+        for backend_tag, replica_tags in self.actor_nursery.replicas.items():
             for replica_tag in replica_tags:
                 replica_name = format_actor_name(replica_tag,
                                                  self.controller_name)
-                self.workers[backend_tag][replica_tag] = ray.get_actor(
-                    replica_name)
+                self.actor_nursery.workers[backend_tag][
+                    replica_tag] = ray.get_actor(replica_name)
 
         # Push configuration state to the router.
         # TODO(edoakes): should we make this a pull-only model for simplicity?
-        for endpoint, traffic_policy in self.traffic_policies.items():
+        for endpoint, traffic_policy in self.configuration_store.\
+                traffic_policies.items():
```

Comment:
The checkpoint recovery logic between the configuration store and the actor nursery seems pretty tightly coupled. I would much prefer if we weren't directly accessing the fields of the actor nursery here (makes it a leaky abstraction). Would it be possible to just do something like load the config from the checkpoint and pass it into the actor nursery and have it reconcile the state & initialize?